packaging/greenboot/microshift-greenboot-required.sh

```diff
@@ -0,0 +1,194 @@
+        echo "The microshift.service systemd unit is failed. Terminating..."
+        kill -TERM ${SCRIPT_PID}
+    fi
+    # Check the service activity
+    [ "${is_active}" = "active" ] && return 0
+    return 1
+}
+
+# Check if any MicroShift pods are in the 'Running' status
+# 
+# args: None
+# return: 0 if any pods are in the 'Running' status, or 1 otherwise
+function any_pods_running() {
+    local count=$(oc get pods ${OCGET_OPT} -A 2>/dev/null | awk '$4~/Running/' | wc -l)
+
+    [ "${count}" -gt 0 ] && return 0
+    return 1
+}
+
+# Check if a given number of MicroShift pods in a given namespace are in the 'Ready' status,
+# terminating the script with the SIGTERM signal if more pods are ready than expected.
+#
+# args: None
+# env1: 'CHECK_PODS_NS' environment variable for the namespace to check
+# env2: 'CHECK_PODS_CT' environment variable for the pod count to check
+# return: 0 if the expected number of pods are ready, or 1 otherwise
+function namespace_pods_ready() {
+    local ns=${CHECK_PODS_NS}
+    local ct=${CHECK_PODS_CT}
+
+    local status=$(oc get pods ${OCGET_OPT} -n ${ns} -o 'jsonpath={..status.conditions[?(@.type=="Ready")].status}' 2>/dev/null)
+    local tcount=$(echo $status | grep -o True  | wc -l)
+    local fcount=$(echo $status | grep -o False | wc -l)
+
+    # Terminate the script in case more pods are ready than expected - nothing to wait for
+    if [ "${tcount}" -gt "${ct}" ] ; then
+        echo "The number of ready pods in the '${ns}' namespace is greater than the expected '${ct}' count. Terminating..."
+        kill -TERM ${SCRIPT_PID}
+    fi
+    # Exit with error if any pods are not ready yet
+    [ "${fcount}" -gt 0 ] && return 1
+    # Check the ready pod count    
+    [ "${tcount}" -eq "${ct}" ] && return 0
+    return 1
+}
+
+# Check if MicroShift pods in a given namespace started and verify they are not restarting by sampling
+# the pod restart count 10 times every 5 seconds and comparing the current sample with the previous one.
+# The pods are considered restarting if the number of 'pod-restarting' samples is greater than the
+# number of 'pod-not-restarting' ones.
+#
+# arg1: Name of the namespace to check
+# return: 0 if pods are not restarting, or 1 otherwise
+function namespace_pods_not_restarting() {
+    local ns=$1
+    local restarts=0
+
+    local count1=$(oc get pods ${OCGET_OPT} -n ${ns} -o 'jsonpath={..status.containerStatuses[].restartCount}' 2>/dev/null)
+    for i in $(seq 10) ; do
+        sleep 5
+        local countS=$(oc get pods ${OCGET_OPT} -n ${ns} -o 'jsonpath={..status.containerStatuses[].started}' 2>/dev/null | grep -vc false)
+        local count2=$(oc get pods ${OCGET_OPT} -n ${ns} -o 'jsonpath={..status.containerStatuses[].restartCount}' 2>/dev/null)
+
+        # If pods started, a restart is detected by comparing the count string between the checks.
+        # The number of pod restarts is incremented when a restart is detected, or decremented otherwise.
```

Comment:
I am not sure if this would work, for example pod restarts are delayed over time (backoff behavior). At some point the delay is bigger than 5secs, the restart counter will keep decrementing.

Also I am not sure about a safe way to check this. For example some pods also restart naturally when you reboot which I guess is what you were trying to avoid detecting as a failure with this behavior...

Reply:
The pod restarts check is performed after making sure pods are ready. If pods are restarting at this point of time, it means there is a problem. That was my thinking at least.

Reply:
But sometimes when you restart pods are reported as ready, even though not even the containers are around yet (checked via crio). ':-)

It's a tricky area.

Reply:
I'm not sure how that's going to happen. This is implemented [here](https://github.com/openshift/microshift/pull/1227/files#diff-dc4eae6d665efc45c9b98bf8a612cbdb19b1fc7e9313affdabd1d158d1d70d8aR93)

Regardless, we're adding image download check, before readiness, so we ensure that all images are there.